### PR TITLE
Add Deeplabv3Plus and DeepLabV3 with segmentation

### PR DESCRIPTION
**Pull request by sachinprasadhs**
Creating the PR from here https://github.com/keras-team/keras-hub/pull/1799 into master branch.

Addressed the previous comments.

Pending:
Need to come up with the preprocessor and resizing for image input as well as masked labels. 
Have to figure out if we can have common preprocessor for segmentation tasks.

### CODE REVIEW

File: keras_hub/src/models/deeplab_v3/deeplab_v3_backbone.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    """DeepLabV3 & DeepLabV3Plus architecture for semantic segmentation.
+
+    This class implements a DeepLabV3 & DeepLabV3Plus architecture as described
+    in [Encoder-Decoder with Atrous Separable Convolution for Semantic Image
+    Segmentation](https://arxiv.org/abs/1802.02611)(ECCV 2018)
+    and [Rethinking Atrous Convolution for Semantic Image Segmentation](
+        https://arxiv.org/abs/1706.05587)(CVPR 2017)
+
+    Args:
+        image_encoder: `keras.Model` instance that is used as a feature
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_backbone.py:34]: add type for args like so 
`image_encoder: `keras.Model`.`
`projection_filter: int.`

File: keras_hub/src/models/deeplab_v3/deeplab_v3_backbone.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        # === Functional Model ===
+        inputs = keras.layers.Input(image_shape)
+
+        fpn_model = keras.Model(
+            image_encoder.inputs, image_encoder.pyramid_outputs
+        )
+
+        fpn_outputs = fpn_model(inputs)
+
+        spatial_pyramid_pooling = SpatialPyramidPooling(
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_backbone.py:103]: define layers on the top 
the general structure is 
```
=== Layers ===
layer1
layer2
.
.
=== Functional Model===
.
.
=== Config ===

File: keras_hub/src/models/deeplab_v3/deeplab_v3_layers.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
+# Copyright 2024 The KerasHUB Authors
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_layers.py:1]: NIT. : `kerasHUB`->`KerasHub`

File: keras_hub/src/models/deeplab_v3/deeplab_v3_layers.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+
+class SpatialPyramidPooling(keras.layers.Layer):
+    """Implements the Atrous Spatial Pyramid Pooling.
+
+    Reference for Atrous Spatial Pyramid Pooling [Rethinking Atrous Convolution
+    for Semantic Image Segmentation](https://arxiv.org/pdf/1706.05587.pdf) and
+    [Encoder-Decoder with Atrous Separable Convolution for Semantic Image
+    Segmentation](https://arxiv.org/pdf/1802.02611.pdf)
+
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_layers.py:26]: add an args sections and an Example section in docstring

File: keras_hub/src/models/deeplab_v3/deeplab_v3_layers.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    inp = keras.layers.Input((384, 384, 3))
+    backbone = keras.applications.EfficientNetB0(
+        input_tensor=inp,
+        include_top=False)
+    output = backbone(inp)
+    output = SpatialPyramidPooling(
+        dilation_rates=[6, 12, 18])(output)
+
+    # output[4].shape = [None, 16, 16, 256]
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_layers.py:35]: remove this comment in docstring

File: keras_hub/src/models/deeplab_v3/deeplab_v3_layers.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        dilation_rates,
+        num_channels=256,
+        activation="relu",
+        dropout=0.0,
+        **kwargs,
+    ):
+        """Initializes an Atrous Spatial Pyramid Pooling layer.
+
+        Args:
+            dilation_rates: A `list` of integers for parallel dilated conv.
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_layers.py:49]: add type for each argument - refer to other classes in KerasHub

File: keras_hub/src/models/image_segmenter.py [hunk 7]
```diff
@@ -27,6 +27,26 @@ class ImageSegmenter(Task):
 
     All `ImageSegmenter` tasks include a `from_preset()` constructor which can
     be used to load a pre-trained config and weights.
+    `Segmenter` tasks take an additional
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/image_segmenter.py:30]: Segmenter->ImageSegmenter

File: keras_hub/src/models/deeplab_v3/deeplab_v3_segmenter.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+    Args:
+        backbone: A `keras_hub.models.DeepLabV3` instance.
+        num_classes: int, the number of classes for the detection model. Note
+            that the `num_classes` contains the background class, and the
+            classes from the data should be represented by integers with range
+            `[0, num_classes]`.
+        activation: str or callable. The activation function to use on
+            the `Dense` layer. Set `activation=None` to return the output
+            logits. Defaults to `None`.
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_segmenter.py:36]: preprocessor missing in args

File: keras_hub/src/models/deeplab_v3/deeplab_v3_layers.py [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import keras
+from keras import ops
+
+
+class SpatialPyramidPooling(keras.layers.Layer):
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_layers.py:19]: please add a test file for `deeplab_v3_layers` and add the [`self.run_layer_tests`](https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/tests/test_case.py#L82) for each of them.

**Review by divyashreepathihalli — COMMENTED** [review]
Thanks for the PR! left some comments.

File: keras_hub/src/models/deeplab_v3/deeplab_v3_backbone.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
+# Copyright 2024 The KerasHUB Authors
```
Comment (fchollet) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_backbone.py:1]: Hub

File: keras_hub/src/models/deeplab_v3/deeplab_v3_backbone.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+)
+
+
+@keras_hub_export("keras_hub.models.DeepLabV3Backbone")
+class DeepLabV3Backbone(Backbone):
+    """DeepLabV3 & DeepLabV3Plus architecture for semantic segmentation.
+
+    This class implements a DeepLabV3 & DeepLabV3Plus architecture as described
+    in [Encoder-Decoder with Atrous Separable Convolution for Semantic Image
+    Segmentation](https://arxiv.org/abs/1802.02611)(ECCV 2018)
```
Comment (fchollet) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_backbone.py:29]: With links, the `[...](` part should be on a single line in order to avoid markdown rendering issues

File: keras_hub/src/models/deeplab_v3/deeplab_v3_backbone.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+    and [Rethinking Atrous Convolution for Semantic Image Segmentation](
```
Comment (fchollet) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_backbone.py:30]: This link is good, for instance!

File: keras_hub/src/models/deeplab_v3/deeplab_v3_backbone.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+        low_level_feature_key: (Optional) str, layer level to extract the feature
+            from one of the key from the `image_encoder`s `pyramid_outputs`
+            property such as  "P2", "P3" etc which will be the Decoder block.
+            Required only when the DeepLabV3Plus architecture needs to be applied.
+        image_shape: tuple. The input shape without the batch size.
+            Defaults to `(None, None, 3)`.
+
+    Example:
+    ```python
+    image_encoder = keras_hub.models.ResNetBackbone.from_preset("resnet_v2_50")
```
Comment (fchollet) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_backbone.py:63]: Add a comment line describing what happens here (loading a ResNet50V2 model with randomly initialized weights?)

File: keras_hub/src/models/deeplab_v3/deeplab_v3_backbone.py [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+            `pyramid_outputs` property such as  "P4", "P5" etc.
+        upsampling_size: Int, or tuple of 2 integers. The upsampling factors for
+            rows and columns of `spatial_pyramid_pooling` layer.
+            If `low_level_feature_key` is given then `spatial_pyramid_pooling`s
+            layer resolution should match with the `low_level_feature`s layer
+            resolution to concatenate both the layers for combined encoder
+            outputs.
+        dilation_rates: A `list` of integers for parallel dilated conv applied to
+        `SpatialPyramidPooling`. Usually a
+            sample choice of rates are [6, 12, 18].
```
Comment (fchollet) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_backbone.py:53]: Use backticks for code keywords

File: keras_hub/src/models/deeplab_v3/deeplab_v3_segmenter.py [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+    Example:
+    ```python
+    images = np.ones(shape=(1, 96, 96, 3))
+    labels = np.zeros(shape=(1, 96, 96, 1))
+    segmenter = keras_hub.models.DeepLabV3ImageSegmenter.from_preset(
```
Comment (fchollet) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_segmenter.py:42]: Add comments describing what the code is doing

File: keras_hub/src/models/deeplab_v3/deeplab_v3_segmenter.py [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+        backbone: A `keras_hub.models.DeepLabV3` instance.
+        num_classes: int, the number of classes for the detection model. Note
+            that the `num_classes` contains the background class, and the
+            classes from the data should be represented by integers with range
+            `[0, num_classes]`.
+        activation: str or callable. The activation function to use on
+            the `Dense` layer. Set `activation=None` to return the output
+            logits. Defaults to `None`.
+
+    Example:
```
Comment (fchollet) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_segmenter.py:38]: More examples may be helpful!

**Comment by divyashreepathihalli** [comment]
@sachinprasadhs can you run `shell/format.sh`

File: keras_hub/src/models/deeplab_v3/deeplab_v3_segmenter_test.py [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+        }
+        self.images = np.ones((2, 96, 96, 3), dtype="float32")
+        self.labels = np.zeros((2, 96, 96, 2), dtype="float32")
+        self.train_data = (
+            self.images,
+            self.labels,
+        )
+
+    def test_classifier_basics(self):
+        pytest.skip(
```
Comment (divyashreepathihalli) [inline keras_hub/src/models/deeplab_v3/deeplab_v3_segmenter_test.py:46]: can you add the preprocessor flow similar to resnet and enable this?

**Review by divyashreepathihalli — APPROVED** [review]
LGTM!
